File: app/src/test/java/cz/civilizacehra/cipherbreaker/StringUtilsTest.kt
package cz.civilizacehra.cipherbreaker

import org.junit.Assert.assertEquals
import org.junit.Test

class StringUtilsTest {

    @Test
    fun levenshteinDistanceOfSameWordsIsZero() {
        assertEquals(0, levenshteinDistance("sifra", "sifra"))
        assertEquals(0, levenshteinDistance("", ""))
    }

    @Test
    fun levenshteinDistanceCountsEdits() {
        assertEquals(3, levenshteinDistance("kitten", "sitting"))
        assertEquals(2, levenshteinDistance("flaw", "lawn"))
        assertEquals(1, levenshteinDistance("sifra", "sifry"))
    }

    @Test
    fun levenshteinDistanceToEmptyWordIsLength() {
        assertEquals(5, levenshteinDistance("sifra", ""))
        assertEquals(5, levenshteinDistance("", "sifra"))
    }

    @Test
    fun levenshteinDistanceIgnoresCase() {
        assertEquals(0, levenshteinDistance("Sifra", "sIFRA"))
    }

    @Test
    fun limitedLevenshteinDistanceIsSameBelowLimit() {
        assertEquals(0, levenshteinDistance("sifra", "sifra", 6))
        assertEquals(3, levenshteinDistance("kitten", "sitting", 6))
        assertEquals(5, levenshteinDistance("sifra", "", 6))
        assertEquals(5, levenshteinDistance("", "sifra", 6))
    }

    @Test
    fun limitedLevenshteinDistanceIsLimitFromLimitUp() {
        assertEquals(3, levenshteinDistance("kitten", "sitting", 3))
        assertEquals(2, levenshteinDistance("kitten", "sitting", 2))
        // Lengths differ by the limit
        assertEquals(6, levenshteinDistance("sifra", "sifrovackou", 6))
        assertEquals(6, levenshteinDistance("abcdefgh", "stuvwxyz", 6))
    }

    @Test
    fun limitedLevenshteinDistanceMatchesUnlimitedOne() {
        val words = listOf("", "a", "kos", "kost", "kosti", "sako", "pesek", "sifra", "sifrovacka",
                "prekvapeni", "prekazka", "nejneobhospodarovavatelnejsi")
        for (a in words) {
            for (b in words) {
                for (limit in 1..8) {
                    assertEquals("$a $b $limit", minOf(levenshteinDistance(a, b), limit), levenshteinDistance(a, b, limit))
                }
            }
        }
    }

    @Test
    fun limitedLevenshteinDistanceDoesNotIgnoreCase() {
        assertEquals(5, levenshteinDistance("Sifra", "sIFRA", 6))
    }

    @Test
    fun levenshteinCostsCanBeReused() {
        val costs = IntArray(20)
        assertEquals(3, levenshteinDistance("kitten", "sitting", 6, costs))
        assertEquals(1, levenshteinDistance("sifra", "sifry", 6, costs))
        assertEquals(6, levenshteinDistance("abcdefgh", "stuvwxyz", 6, costs))
        assertEquals(2, levenshteinDistance("flaw", "lawn", 6, costs))
    }

    @Test
    fun hammingDistanceCountsDifferentPositions() {
        assertEquals(0, hammingDistance("sifra", "sifra"))
        assertEquals(3, hammingDistance("karolin", "kathrin"))
        assertEquals(5, hammingDistance("abcde", "vwxyz"))
    }

    @Test
    fun hammingDistanceIgnoresCase() {
        assertEquals(0, hammingDistance("Sifra", "sIFRA"))
    }
}

File: pwa/test/string-utils.test.ts
import { describe, expect, it } from 'vitest';
import { hammingDistance, levenshteinDistance } from '../src/logic/string-utils';

describe('levenshteinDistance', () => {
  it('is zero for the same words', () => {
    expect(levenshteinDistance('sifra', 'sifra')).toBe(0);
    expect(levenshteinDistance('', '')).toBe(0);
  });

  it('counts edits', () => {
    expect(levenshteinDistance('kitten', 'sitting')).toBe(3);
    expect(levenshteinDistance('flaw', 'lawn')).toBe(2);
    expect(levenshteinDistance('sifra', 'sifry')).toBe(1);
  });

  it('is the length to an empty word', () => {
    expect(levenshteinDistance('sifra', '')).toBe(5);
    expect(levenshteinDistance('', 'sifra')).toBe(5);
  });

  it('ignores case', () => {
    expect(levenshteinDistance('Sifra', 'sIFRA')).toBe(0);
  });

  it('is the same below the limit', () => {
    expect(levenshteinDistance('sifra', 'sifra', 6)).toBe(0);
    expect(levenshteinDistance('kitten', 'sitting', 6)).toBe(3);
    expect(levenshteinDistance('sifra', '', 6)).toBe(5);
    expect(levenshteinDistance('', 'sifra', 6)).toBe(5);
  });

  it('is the limit from the limit up', () => {
    expect(levenshteinDistance('kitten', 'sitting', 3)).toBe(3);
    expect(levenshteinDistance('kitten', 'sitting', 2)).toBe(2);
    // Lengths differ by the limit
    expect(levenshteinDistance('sifra', 'sifrovackou', 6)).toBe(6);
    expect(levenshteinDistance('abcdefgh', 'stuvwxyz', 6)).toBe(6);
  });

  it('matches the unlimited one when limited', () => {
    const words = ['', 'a', 'kos', 'kost', 'kosti', 'sako', 'pesek', 'sifra', 'sifrovacka',
      'prekvapeni', 'prekazka', 'nejneobhospodarovavatelnejsi'];
    for (const a of words) {
      for (const b of words) {
        for (let limit = 1; limit <= 8; limit++) {
          expect(levenshteinDistance(a, b, limit), `${a} ${b} ${limit}`).toBe(Math.min(levenshteinDistance(a, b), limit));
        }
      }
    }
  });

  it('does not ignore case when limited', () => {
    expect(levenshteinDistance('Sifra', 'sIFRA', 6)).toBe(5);
  });

  it('can reuse the costs', () => {
    const costs = new Int32Array(20);
    expect(levenshteinDistance('kitten', 'sitting', 6, costs)).toBe(3);
    expect(levenshteinDistance('sifra', 'sifry', 6, costs)).toBe(1);
    expect(levenshteinDistance('abcdefgh', 'stuvwxyz', 6, costs)).toBe(6);
    expect(levenshteinDistance('flaw', 'lawn', 6, costs)).toBe(2);
  });
});

describe('hammingDistance', () => {
  it('counts different positions', () => {
    expect(hammingDistance('sifra', 'sifra')).toBe(0);
    expect(hammingDistance('karolin', 'kathrin')).toBe(3);
    expect(hammingDistance('abcde', 'vwxyz')).toBe(5);
  });

  it('ignores case', () => {
    expect(hammingDistance('Sifra', 'sIFRA')).toBe(0);
  });
});
